### PR TITLE
Add documentation to the API

### PR DESCRIPTION
**Pull request by vibhoothiiaanand**
This commit Adds documentation to multiple places
Fixes the follwing issues
- [ ] #1075 
- [ ] #1076 
- [ ] #1077 
- [ ] #1078 
- [ ] #1079 

I Created a single PR so it will be easy for reviewing and closing rather than 5 different PRs for 5 issues.

### CODE REVIEW

File: src/api.rs [hunk 1]
```diff
@@ -89,6 +93,10 @@ impl Default for EncoderConfig {
 }
 
 impl EncoderConfig {
+  /// This is the preset set with width and height to 640 and 480 respectively,
```
Comment (lu-zero) [inline src/api.rs]: Maybe omit the details that could and will change and just mention that provides a default setting according to a speed value in the range `0-10`.

Mention what happens when you pass a number higher than `10`

File: src/api.rs [hunk 2]
```diff
@@ -81,6 +82,9 @@ pub struct EncoderConfig {
   pub train_rdo: bool,
 }
 
+/// Default preset for EncoderConfig, In this DEFAULT_SPEED is set to 5 
+/// and called the with_speed_preset with this value(5). We are using 5 
+/// as it is in the middle so it will have good balance of quality and speed.
```
Comment (lu-zero) [inline src/api.rs]: Omit internal information that might or might not change, just mention that the default is a balance between speed and quality and mention that the user could use `with_speed_preset` to have a default structure with a specific speed level.

File: src/api.rs [hunk 3]
```diff
@@ -154,6 +164,9 @@ impl Default for SpeedSettings {
 }
 
 impl SpeedSettings {
+  /// This preset is set to assign the speed, where the speed is passed as arguement,
+  /// the parametes like min_block_size, multi reference preditions, distortion are assigned,
+  /// for instance  *speed_settings: SpeedSettings::from_preset(10);*
```
Comment (lu-zero) [inline src/api.rs]: "Set the speed setting according to a numeric speed preset"

Mention which is the range and what's slower/higher quality and what's faster.

File: src/api.rs [hunk 4]
```diff
@@ -282,14 +296,16 @@ pub enum ChromaSampling {
   Cs400,
 }
 
+/// Make 4:2:0 as default chroma sampling type as it is almost loseless since it
+/// is removing half of the horizondtal and half of vertical color information
```
Comment (lu-zero) [inline src/api.rs]: It is not lossless. You may omit the default description.

File: src/api.rs [hunk 5]
```diff
@@ -309,13 +326,15 @@ pub enum ChromaSamplePosition {
...
 
+/// Assign default ChromaSamplePosition as Unknown
 impl Default for ChromaSamplePosition {
   fn default() -> Self {
     ChromaSamplePosition::Unknown
   }
 }
 
 arg_enum!{
+  /// Contains details of Pixel color range
```
Comment (lu-zero) [inline src/api.rs]: It does not add more information, you may add a reference to what is a color range is and the ITU H.273 section that mentions it.

File: src/api.rs [hunk 6]
```diff
@@ -325,13 +344,15 @@ arg_enum!{
   }
 }
 
+/// Assign default of PixelRange as Unspecified
```
Comment (lu-zero) [inline src/api.rs]: Omit it.

File: src/api.rs [hunk 7]
```diff
@@ -325,13 +344,15 @@ arg_enum!{
...
 impl Default for PixelRange {
     fn default() -> Self {
         PixelRange::Unspecified
     }
 }
 
 arg_enum!{
+  /// Contains different types of Matrix Coefficients which is used in deriving the luma and chroma (YCbCr) from the R'G'B' primaries
```
Comment (lu-zero) [inline src/api.rs]: A reference to ITU H.273 would be good.

File: src/api.rs [hunk 8]
```diff
@@ -352,13 +373,15 @@ arg_enum!{
   }
 }
 
+/// Assign default of MatrixCoefficients as Unspecified
```
Comment (lu-zero) [inline src/api.rs]: Omit it.

File: src/api.rs [hunk 9]
```diff
@@ -377,13 +400,15 @@ arg_enum!{
   }
 }
 
+/// Assign default of ColorPrimaries as Unspecified
```
Comment (lu-zero) [inline src/api.rs]: Omit it.

File: src/api.rs [hunk 10]
```diff
@@ -377,13 +400,15 @@ arg_enum!{
...
 impl Default for ColorPrimaries {
     fn default() -> Self {
         ColorPrimaries::Unspecified
     }
 }
 
 arg_enum!{
+  /// Contains differnt types of Transfer Characteristics, they define the optoelectronic tranfer charecteristic of the source picture aka Gamma.
```
Comment (lu-zero) [inline src/api.rs]: Typo here -> `different`.

(mention to the H.273 section would be great)

File: src/api.rs [hunk 11]
```diff
@@ -407,19 +432,23 @@ arg_enum!{
   }
 }
 
+/// Assign default of TransferCharacteristics as Unspecified
```
Comment (lu-zero) [inline src/api.rs]: Feel free to omit this.

File: src/api.rs [hunk 12]
```diff
@@ -445,6 +476,7 @@ pub struct Config {
 const MAX_USABLE_THREADS: usize = 4;
 
 impl Config {
+  /// Contain the values of the encoder configuration which are to be parsed
```
Comment (lu-zero) [inline src/api.rs]: "Sets a specific configuration `key` parameter by parsing the `value` string"

File: src/api.rs [hunk 13]
```diff
@@ -459,11 +491,13 @@ impl Config {
 
     Ok(())
   }
-
+  /// Create a new context for the encoder. Context contatins
```
Comment (lu-zero) [inline src/api.rs]: The context is the full encoder state.

Please omit implementation details that may and will change and mention what the context is useful for :)

File: src/api.rs [hunk 14]
```diff
@@ -477,7 +511,7 @@ impl Config {
     };
 
     let pool = rayon::ThreadPoolBuilder::new().num_threads(threads).build().unwrap();
-
+    // Initialize default values to context objects
```
Comment (lu-zero) [inline src/api.rs]: May be omitted.

File: src/api.rs [hunk 15]
```diff
@@ -509,6 +543,7 @@ impl Config {
   }
 }
 
+/// Contain all the context configurations
```
Comment (lu-zero) [inline src/api.rs]: Omit it or expand it.

File: src/api.rs [hunk 16]
```diff
@@ -570,6 +605,8 @@ impl<T: Pixel> fmt::Display for Packet<T> {
 }
 
 impl<T: Pixel> Context<T> {
+
+  /// Create new frame  with details like width, height and also chroma sampling from the encoder
```
Comment (lu-zero) [inline src/api.rs]: "Create a new empty frame according to the current encoder settings"

File: src/api.rs [hunk 17]
```diff
@@ -578,6 +615,7 @@ impl<T: Pixel> Context<T> {
     ))
   }
 
+  /// Send the details of the frame by passing the frame details as input
```
Comment (lu-zero) [inline src/api.rs]: Send a frame to be encoded.

Mention that the function may return an error if the input queue is full.

File: src/api.rs [hunk 18]
```diff
@@ -589,32 +627,39 @@ impl<T: Pixel> Context<T> {
     Ok(())
   }
 
+  /// Returns the details of a specific frame
```
Comment (lu-zero) [inline src/api.rs]: You may omit this, (and this function should not be public btw).

File: src/api.rs [hunk 19]
```diff
@@ -589,32 +627,39 @@ impl<T: Pixel> Context<T> {
...
   pub fn get_frame(&self, frame_number: u64) -> Arc<Frame<T>> {
     // Clones only the arc, so low cost overhead
     self.frame_q.get(&frame_number).as_ref().unwrap().as_ref().unwrap().clone()
   }
 
+  /// Returns the frame count
```
Comment (lu-zero) [inline src/api.rs]: You should explain what the frame count is.

File: src/api.rs [hunk 20]
```diff
@@ -589,32 +627,39 @@ impl<T: Pixel> Context<T> {
...
   pub fn get_frame_count(&self) -> u64 {
     self.frame_count
   }
 
+  /// Set the limit which is passed as argument
```
Comment (lu-zero) [inline src/api.rs]: You should explain what the limit is.

File: src/api.rs [hunk 21]
```diff
@@ -589,32 +627,39 @@ impl<T: Pixel> Context<T> {
...
   pub fn set_limit(&mut self, limit: u64) {
     self.limit = limit;
   }
 
+  /// Allow to read frames ahead of current ones
```
Comment (lu-zero) [inline src/api.rs]: This function should not be public. (and does not do that)

File: src/api.rs [hunk 22]
```diff
@@ -589,32 +627,39 @@ impl<T: Pixel> Context<T> {
...
   pub fn needs_more_lookahead(&self) -> bool {
     self.needs_more_frames(self.frame_count) && self.frames_processed + LOOKAHEAD_FRAMES > self.frame_q.keys().last().cloned().unwrap_or(0)
   }
 
+  /// Request for more frames
```
Comment (lu-zero) [inline src/api.rs]: Returns true if more frames should be sent to the encoder in order.

Mention `send_frame()`.

File: src/api.rs [hunk 23]
```diff
@@ -589,32 +627,39 @@ impl<T: Pixel> Context<T> {
...
   pub fn needs_more_frames(&self, frame_count: u64) -> bool {
     self.limit == 0 || frame_count < self.limit
   }
 
+  /// Video sequences are headed by a sequence header which was starting value, various other things like chroma sampling x and y, sequence bit depth
```
Comment (lu-zero) [inline src/api.rs]: You should document what the function returns. Bonus point if you add a reference to the AV1 specification.

File: src/api.rs [hunk 24]
```diff
@@ -741,6 +786,10 @@ impl<T: Pixel> Context<T> {
     (fi, true)
   }
 
+  /// Receive packet functions provide an Encode API with decoupled packets/frames, On sucess it returns packet with compressed frame.
```
Comment (lu-zero) [inline src/api.rs]: Do not mention implementation details that may and should be changed soon :)

"Returns an encoded packet or signals that additional frame data must be provided"

Mention `flush()` and `send_frame()` .

File: src/api.rs [hunk 25]
```diff
@@ -876,6 +925,9 @@ impl<T: Pixel> Context<T> {
     }
   }
 
+  /// Flush the output, the flush makes the current frame to the frame_q and then increments the frame count for the next process.
```
Comment (lu-zero) [inline src/api.rs]: Please omit internal details.

Mention `receive_packet()` and `send_frame()`.

File: src/api.rs [hunk 26]
```diff
@@ -48,6 +48,7 @@ pub struct Point {
   pub y: u16
 }
 
+/// Contain all the encoder configuration
```
Comment (shssoichiro) [inline src/api.rs]: Can you add that this should contain settings which affect the encoded output? Any options which do not affect the output should be in `Config` instead.

Comment (vibhoothiiaanand) [reply src/api.rs]: Sure :) 
Updated

Comment (vibhoothiiaanand) [reply src/api.rs]: Done

Comment (vibhoothiiaanand) [reply src/api.rs]:
```


/// Contain all the encoder configuration which are taken for encoded output.
/// Encoder configurations which doesn't affect the encoded output are in 'config'.
#[derive(Clone, Debug)]
```

File: src/api.rs (hunk 15 above)
Comment (vibhoothiiaanand) [reply src/api.rs]: Omitting it as the function has explanation below

**Comment by rzumer** [comment]
Did you forget to push changes?

**Comment by vibhoothiiaanand** [comment]
> Did you forget to push changes? I don't see the fixes that should be there.

I didn't push, was bit busy with Midterms and project proposal and also was waiting for Luca to fix up the issues of cargo doc. Now all good, Will update few more then rebase and will push :)

**Comment by lu-zero** [comment]
Please open a new pr with a single patch addressing a single issue. (Possibly not touching the Context since is being actively reworked)

**Pull request closed without merging**